File: app/lib/ops.ts
import { GraphQLClient } from 'graphql-request';
import fetch from 'node-fetch';
import https from 'https';
import fs from 'fs';

interface Application {
  name: string;
  subdomain: string;
  status: string;
}

interface ListApplicationsResponse {
  applications: Application[];
}

/**
 * Get a global trace ID from a local hex trace ID. Phoenix GraphQL does not
 * allow to search for projects by a project ID or by project name, but the
 * endpoint DOES allow us to get individual nodes by their "global ID". We use
 * this query to get the global ID of the trace whose "local" ID (predefined
 * as the actual trace ID of the span context) matches the trace ID returned
 * when kicking off a crew.
 */
const getProjectAndTraceInfo = async (client: GraphQLClient, traceId: string) => {
  const query = `
query QueryProjectsForTraceExistence {
  projects(first: 1000) {
    edges {
      node {
        id
        trace(traceId: "${traceId}") {
          id
        }
      }
    }
  }
}
  `;
  const data: any = await client.request(query);
  const project: any = data.projects.edges.find((edge: any) => edge.node.trace?.id);
  const globalTraceId: string = project.node.trace.id;
  return {
    projectId: project.node.id,
    globalTraceId: globalTraceId,
  };
};

const eventTypes = [
  'Crew.kickoff',
  'Agent._start_task',
  'completion',
  'ToolUsage._use',
  'ToolUsage._end_use',
  'Agent._end_task',
  'Crew.complete',
];

/**
 * Get all crew events given a specific crew Trace. It's assumed that the
 * traceId is the "local" trace ID that was passed from the crew kickoff call.
 */
export const getCrewEvents = async (client: GraphQLClient, traceId: string) => {
  // Get the global trace ID
  const { projectId, globalTraceId } = await getProjectAndTraceInfo(client, traceId);

  // Query the global trace ID
  const query = `
query MyQuery {
  node(id: "${globalTraceId}") {
    ... on Trace {
      rootSpan {
        name
        descendants {
          id
          name
          startTime
          cumulativeTokenCountTotal
          cumulativeTokenCountPrompt
          cumulativeTokenCountCompletion
          endTime
          attributes
          events {
            message
            name
            timestamp
          }
        }
      }
    }
  }
}
  `;
  const data: any = await client.request(query);
  const events: any[] = [];
  data.node.rootSpan.descendants.map((descendant: any) => {
    if (eventTypes.includes(descendant.name)) {
      events.push({
        ...descendant,
        attributes: JSON.parse(descendant.attributes),
        events: descendant.events || [],
      });
    }
    return;
  });

  // Sort the data by startTime in ascending order
  const sortedEvents = events.sort((a: any, b: any) => {
    const dateA = new Date(a.startTime);
    const dateB = new Date(b.startTime);
    return dateA.getTime() - dateB.getTime(); // Compare timestamps
  });

  return {
    projectId: projectId,
    events: sortedEvents,
  };
};

export const fetchOpsUrl = async (): Promise<string | null> => {
  const CDSW_APIV2_KEY = process.env.CDSW_APIV2_KEY;
  const CDSW_DOMAIN = process.env.CDSW_DOMAIN;
  const CDSW_PROJECT_ID = process.env.CDSW_PROJECT_ID;

  if (process.env.AGENT_STUDIO_DEPLOYMENT_CONFIG === 'dev') {
    return 'http://127.0.0.1:8123';
  }

  if (!CDSW_APIV2_KEY || !CDSW_DOMAIN || !CDSW_PROJECT_ID) {
    console.error('Environment variables are not set properly.');
    return null;
  }

  // Use the CA bundle
  const agent = new https.Agent({
    ca: fs.readFileSync('/etc/ssl/certs/ca-certificates.crt'),
  });

  try {
    const response = await fetch(
      `https://${CDSW_DOMAIN}/api/v2/projects/${CDSW_PROJECT_ID}/applications`,
      {
        headers: {
          authorization: `Bearer ${CDSW_APIV2_KEY}`,
        },
        agent,
      },
    );

    if (!response.ok) {
      throw new Error(`Failed to fetch applications: ${response.statusText}`);
    }

    const data = (await response.json()) as ListApplicationsResponse;

    // Find the most recent running version of Agent Studio
    const runningApps = data.applications.filter(
      (app: { name: string; status: string }) =>
        app.name.toLowerCase().includes('agent studio - agent ops & metrics') &&
        app.status.toLowerCase().includes('running'),
    );

    if (runningApps.length === 0) {
      console.error("No running 'Agent Studio - Agent Ops & Metrics' applications found.");
      return null;
    }

    // Sort by version if present (assuming format "Name vX.Y")
    const getVersion = (appName: string): number[] => {
      try {
        const version = appName.split('v').pop() || '';
        return version.split('.').map(Number);
      } catch {
        return [0, 0]; // Default for apps without version
      }
    };

    // Get the most recent version
    const application = runningApps.sort((a, b) => {
      const vA = getVersion(a.name);
      const vB = getVersion(b.name);
      return vB[0] - vA[0] || vB[1] - vA[1];
    })[0];

    if (!application) {
      console.error("No suitable 'Agent Studio - Agent Ops & Metrics' application found.");
      return null;
    }

    const outputURL = `https://${application.subdomain}.${CDSW_DOMAIN}`;
    return outputURL;
  } catch (error) {
    console.error('Error fetching applications:', error);
    return null;
  }
};

File: studio/sdk/ops.py
import json
from gql import gql, Client
from datetime import datetime

# This is the Python equivalent of your TypeScript constants:
EVENT_TYPES = [
    "Crew.kickoff",
    "Agent._start_task",
    "completion",
    "ToolUsage._use",
    "ToolUsage._end_use",
    "Agent._end_task",
    "Crew.complete",
]


def get_project_and_trace_info(client: Client, local_trace_id: str):
    """
    Given a local trace ID, query all projects and find the one whose node.trace.id exists.
    Returns a dictionary containing { "projectId": ..., "globalTraceId": ... }.
    """

    # GraphQL query in a multi-line string.
    # We'll embed the `trace_id` into the query string:
    query_str = f"""
    query QueryProjectsForTraceExistence {{
      projects(first: 1000) {{
        edges {{
          node {{
            id
            trace(traceId: "{local_trace_id}") {{
              id
            }}
          }}
        }}
      }}
    }}
    """

    # Structure your payload:
    result = client.execute(gql(query_str))

    edges = result["projects"]["edges"]

    # Find the project whose 'trace' is not null and has an 'id'
    project = next((edge for edge in edges if edge["node"]["trace"] and edge["node"]["trace"].get("id")), None)

    if not project:
        raise ValueError(f"No project found for traceId={local_trace_id}")

    global_trace_id = project["node"]["trace"]["id"]
    project_id = project["node"]["id"]

    return {"projectId": project_id, "globalTraceId": global_trace_id}


def get_crew_events(client: Client, local_trace_id: str) -> dict:
    """
    Get all "descendants" spanning events for the global trace that corresponds
    to a local trace ID. Returns a dict with keys "projectId" and "events".
    """
    # First, retrieve the global trace ID (and project ID) from your helper:
    project_and_trace_info = get_project_and_trace_info(client, local_trace_id)
    project_id = project_and_trace_info["projectId"]
    global_trace_id = project_and_trace_info["globalTraceId"]

    # GraphQL query using a variable for the global trace ID
    query_str = f"""
    query GetCrewEvents {{
        node(id: "{global_trace_id}") {{
            ... on Trace {{
            rootSpan {{
                name
                descendants {{
                id
                name
                startTime
                cumulativeTokenCountTotal
                cumulativeTokenCountPrompt
                cumulativeTokenCountCompletion
                endTime
                attributes
                events {{
                    message
                    name
                    timestamp
                }}
                }}
            }}
            }}
        }}
    }}
    """

    # Execute the query with the global trace ID
    result = client.execute(gql(query_str))

    node_data = result["node"]
    if not node_data or not node_data.get("rootSpan"):
        return {"projectId": project_id, "events": []}

    descendants = node_data["rootSpan"]["descendants"] or []

    # Filter to the spans whose `name` is in EVENT_TYPES
    filtered_events = []
    for d in descendants:
        if d["name"] in EVENT_TYPES:
            # Parse the "attributes" JSON, if present
            parsed_attributes = {}
            if d["attributes"]:
                try:
                    parsed_attributes = json.loads(d["attributes"])
                except json.JSONDecodeError:
                    parsed_attributes = {}

            event_object = {
                **d,
                "attributes": parsed_attributes,
                "events": d.get("events", []),
            }
            filtered_events.append(event_object)

    # Sort by the 'startTime' ascending, converting ISO string to datetime
    def parse_time(e):
        # e.g. '2023-03-12T12:34:56.789Z'
        return datetime.fromisoformat(e["startTime"].replace("Z", "+00:00"))

    filtered_events.sort(key=parse_time)

    return {"projectId": project_id, "events": filtered_events}
